runtime/main/src/main/java/org/qbicc/runtime/main/VMHelpers.java

```diff
@@ -111,48 +104,71 @@ private static boolean isAssignableToLeaf(type_id valueTypeId, type_id toTypeId)
     private static void omError(c_int nativeErrorCode) throws IllegalMonitorStateException {
         int errorCode = nativeErrorCode.intValue();
         if (0 != errorCode) {
-            throw new IllegalMonitorStateException("error code is: " + errorCode);
+//            putchar('O');
+//            putchar('M');
+//            putchar('E');
+//            putchar('R');
+//            putchar('R');
+//            putchar(':');
+//            printInt(errorCode);
+            throw new IllegalMonitorStateException("error code: " + errorCode);
         }
     }
 
     // TODO: mark this with a "NoInline" annotation
     static void monitor_enter(Object object) throws IllegalMonitorStateException {
-        /* TODO: deal with racy nature of this creation */
-        if (objectMonitorNatives == null) {
-            objectMonitorNatives = new ConcurrentHashMap<>();
-        }
-
-        // TODO malloc(sizeof(class)) resulted in "invalid coercion of s64 to u64" this is a workaround
-        Stddef.size_t mutexAttrSize = sizeof(pthread_mutexattr_t.class);
-        ptr<?> attrVoid = malloc(word(mutexAttrSize.longValue()));
-        if (attrVoid.isNull()) {
-            throw new OutOfMemoryError(/*"Allocation failed"*/);
+        if (object == null) {
+            /* TODO skip for now. Object should never be null except that
+                classof_from_typeid is not currently implemented. */
+            return;
         }
-        ptr<pthread_mutexattr_t> attr = (ptr<pthread_mutexattr_t>)castPtr(attrVoid, pthread_mutexattr_t.class);
+        while (true) { /* facilitate restart if atomic set operation fails */
+            pthread_mutex_t_ptr nom = ObjectModel.nativeObjectMonitor_of(object);
+            if (nom == null) {
+                ptr<?> attrVoid = malloc(sizeof(pthread_mutexattr_t.class));
+                if (attrVoid.isNull()) {
+                    throw new OutOfMemoryError(/*"Allocation failed"*/);
+                }
+                ptr<pthread_mutexattr_t> attr = (ptr<pthread_mutexattr_t>) castPtr(attrVoid, pthread_mutexattr_t.class);
 
-        Stddef.size_t mutexSize = sizeof(pthread_mutex_t.class);
-        ptr<?> mVoid = malloc(word(mutexSize.longValue()));
-        if (attrVoid.isNull()) {
-            throw new OutOfMemoryError(/*"Allocation failed"*/);
+                Stddef.size_t mutexSize = sizeof(pthread_mutex_t.class);
+                nom = malloc(word(mutexSize.longValue()));
+                ptr<?> mVoid = malloc(word(mutexSize.longValue()));
```

Comment:
Why are there two mallocs here?  Only one is null checked and it looks like the one that isn't checked is always leaked

Reply:
thanks. forgot to remove this